docs/design/1600-prometheus-relabel-external-cache.md

```diff
@@ -0,0 +1,172 @@
+# Proposal: Prometheus relabel external cache
+
+- Author(s): Wilfried ROSET (@wilfriedroset), Pierre BAILHACHE (@pbailhache)
+- Last updated: 2024-09-02
+- Original issue: <https://github.com/grafana/alloy/issues/1600>
+
+## Abstract
+
+This proposal introduces a way to configure the component `prometheus.relabel` so that it could use an external cache such as `Redis` or `Memcached` instead of `in-memory`.
+
+## Problem
+
+The `prometheus.relabel` component rewrites the label set of each metric passed along to the exported receiver by applying one or more relabeling rules. To do so it uses a relabeling cache stored in memory. For each `prometheus.relabel` component Alloy will create a dedicated relabel cache. This is not a huge issue per se due to the possibility of registering multiple rules into on component to mutualize the cache.
+
+However if you're horizontally scaling Alloy deployment with load-balancing, you will end up with one cache per Alloy instance and those local caches will have overlaps, increasing the footprint of each instance.
+Moreover, the cache is tied to the pod, which means that a new Alloy process starts with an empty cache.
+
+For a couple of horizontally scaled Alloy pods it's acceptable but if you plan to have lots of instances processing data horizontally it's not sustainable.
+
+## Proposal
+
+Allow to use `Redis` or `Memcached` instead of the `in-memory` one as cache.
+
+Using [dskit](https://github.com/grafana/dskit/blob/main/cache/cache.go) code to manage connection and client configuration.
+
+We could create an interface to avoid changing the code logic and to abstract the kind of cache used to the component.
+
+## Pros and cons
+
+**Pros:**
+
+- No logic change so impact is expected to be null for users
+- Possibility to use an external cache if needed, even having multiple caches for different relabeling components.
+- Will be easy to use in other parts of the codebase
+
+**Cons:**
+
+- Config is a bit more complex compared to previous one
+
+## Alternative solutions
+
+The alternative is to do nothing as we deem this improvement unnecessary.
+
+## Compatibility
+
+This proposal offer to deprecate for a couple of release the old way to configure the in-memory cache, then release+2 drop the old way. Doing so allow to migrate the settings to the correct block.
```

Comment:
I think our accepted [backwards compatibility](https://grafana.com/docs/alloy/latest/introduction/backward-compatibility/) guarantees go against this. Any breaking changes for stable functionality would have to wait until v2.0.

Reply:
We don't want to deprecate anything sooner than what is guaranteed.
I will rephrase to say that we keep both compatibility until the next major release